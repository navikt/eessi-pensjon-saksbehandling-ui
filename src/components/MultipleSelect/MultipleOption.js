import React from 'react'
import PT from 'prop-types'
import classNames from 'classnames'
import { Checkbox } from 'components/Nav'

const MultipleOption = (props) => {
  const { data, selectProps, innerProps, isSelected, isFocused } = props

  return <div id={selectProps.id + '-' + data.value}>
    <div className={classNames('c-multipleOption', {
      selected: isSelected,
      focused: isFocused
    })} {...innerProps}>
      <Checkbox
        className='c-multipleOption__checkbox'
        label={data.label}
        onChange={() => {}} checked={isSelected}
      />
    </div>
  </div>
}

MultipleOption.propTypes = {
  data: PT.object,
  selectProps: PT.object,
  innerProps: PT.object,
  isSelected: PT.bool,
  isFocused: PT.bool
}

export default MultipleOption
