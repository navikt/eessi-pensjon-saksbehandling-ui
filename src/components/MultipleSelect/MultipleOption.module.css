.label:hover {
  color: var(--a-text-on-inverted) !important;
}

