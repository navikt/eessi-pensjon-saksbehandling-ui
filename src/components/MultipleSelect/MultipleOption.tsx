import { themeKeys } from 'nav-styled-component-theme'
import PT from 'prop-types'
import React from 'react'
import { OptionProps } from 'react-select'
import { Checkbox } from 'nav-frontend-skjema'
import styled, { ThemeProvider } from 'styled-components'

export type MultipleOptionProps = OptionProps<any>

const OptionCheckbox = styled(Checkbox)`
   &:not(:disabled) {
     cursor: pointer
   }
   label {
      &:hover {
       color: ${({ theme }) => theme[themeKeys.INVERTED_FONT_COLOR]} !important;
     }
   }
`

const MultipleOption: React.FC<MultipleOptionProps> = (props: MultipleOptionProps): JSX.Element => {
  const { data, selectProps, innerProps, isSelected, getStyles } = props
  const id: string = selectProps.id + '-' + data.value
  const theme = selectProps.selectProps.theme

  return (
    <div
      {...innerProps}
      style={getStyles('option', props)}
      onClick={(e) => {
        e.preventDefault()
        e.stopPropagation()
        innerProps.onClick(e)
      }}
    >
      <ThemeProvider theme={theme}>
        <OptionCheckbox
          data-test-id={'c-multipleOption__checkbox-' + id}
          label={data.label}
          onChange={(e) => {
            e.preventDefault()
            e.stopPropagation()
          }}
          checked={isSelected}
        />
      </ThemeProvider>
    </div>
  )
}

MultipleOption.propTypes = {
  data: PT.object,
  selectProps: PT.object.isRequired,
  innerProps: PT.any.isRequired,
  isSelected: PT.bool.isRequired,
  isFocused: PT.bool.isRequired
}
MultipleOption.displayName = 'MultipleOption'
export default MultipleOption
