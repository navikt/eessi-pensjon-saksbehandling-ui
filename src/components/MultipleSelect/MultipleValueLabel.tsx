import { components } from 'react-select'
import styled from 'styled-components'

const Label = styled.div`
  padding: 0px;
`
const FlexDiv = styled.div`
  display: flex;
  align-items: center;
`
const MultipleValueLabel = (props: any) => (
  <components.MultiValueLabel data-test-id='c-multipleselect-multivaluelabel' {...props}>
    <FlexDiv>
      <Label>&nbsp;&nbsp;&nbsp;&nbsp;{props.data.label}&nbsp;&nbsp;&nbsp;&nbsp;</Label>
    </FlexDiv>
  </components.MultiValueLabel>
)

export default MultipleValueLabel
