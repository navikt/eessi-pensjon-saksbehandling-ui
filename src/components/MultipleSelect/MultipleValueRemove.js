import React from 'react'
import Icons from 'components/Icons'
import { components } from 'react-select'
const MultipleValueRemove = (props) => {
  return <components.MultiValueRemove {...props} className='c-multipleSelect-multipleValueRemove'>
    <Icons kind='close' />
  </components.MultiValueRemove>
}

export default MultipleValueRemove
