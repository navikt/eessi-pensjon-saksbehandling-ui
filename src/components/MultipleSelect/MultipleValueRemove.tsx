import React, { useState } from 'react'
import FilledRemoveCircle from 'assets/icons/filled-version-remove-circle'
import styled from 'styled-components'

const FlexDiv = styled.div`
  display: flex;
  cursor: pointer;
  &:hover {
    background-color: transparent !important;
  }
`
const MultipleValueRemove = (props: any) => {
  const [hover, setHover] = useState<boolean>(false)
  const { selectProps, innerProps } = props
  const theme = selectProps.selectProps.theme

  const visibility = selectProps.isDisabled ? 'hidden' : 'visible'
  return (
    <FlexDiv
      {...innerProps}
      style={{}}
      data-test-id='c-multipleselect-multiplevalueremove'
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}

    >
      <FilledRemoveCircle
        style={{ visibility: visibility }}
        color={
        theme[hover ? 'main-interactive-color' : 'main-font-color']
      }
      />
    </FlexDiv>

  )
}
export default MultipleValueRemove
