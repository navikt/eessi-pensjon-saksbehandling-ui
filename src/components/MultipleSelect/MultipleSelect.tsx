import classNames from 'classnames'
import MultipleValueLabel from 'components/MultipleSelect/MultipleValueLabel'
import MultipleValueRemove from 'components/MultipleSelect/MultipleValueRemove'
import _ from 'lodash'
import { guid } from 'nav-frontend-js-utils'
import PT from 'prop-types'
import Select, { OptionsOrGroups, GroupBase, OnChangeValue, PropsValue } from 'react-select'
import CreatableSelect from 'react-select/creatable'
import styled from 'styled-components'
import MultipleOption from './MultipleOption'
import { Option } from 'declarations/app'

const MultipleSelectDiv = styled.div`
  .skjemaelement__feilmelding {
    .multipleSelect__control {
      border: 1px solid ${({ theme }) => theme.redError};
    }
  }
`

export interface MultipleSelectProps<T>{
  ariaLabel ?: string
  className ?: string
  creatable ?: boolean
  error ?: string
  hideSelectedOptions ?: boolean
  id ?: string
  isDisabled ?: boolean
  isLoading?: boolean
  isSearchable ?: boolean
  label: string | JSX.Element
  menuPortalTarget?: any,
  onSelect?: (e: Array<T>) => void
  options?: OptionsOrGroups<T, GroupBase<T>>
  size ?: 'medium' | 'small'
  values?: PropsValue<T>
}

const MultipleSelect = <T extends Option = Option> ({
  ariaLabel, className, creatable = false, error, hideSelectedOptions = false, id,
  isDisabled = false, isLoading = false, isSearchable = true, label, menuPortalTarget, onSelect,
  options = [], size = 'medium', values = []
}: MultipleSelectProps<T>): JSX.Element => {
  const onSelectChange = (e: OnChangeValue<T, true>) => {
    if (_.isFunction(onSelect)) {
      onSelect(e as Array<T>)
    }
  }

  const Component = creatable ? CreatableSelect : Select
  const inputId = id || guid()

  const customProps: any = {
    id
  }

  return (
    <MultipleSelectDiv
      className={classNames(className, { skjemaelement__feilmelding: error })}
    >
      {label && (<label className='navds-text-field__label navds-label' htmlFor={inputId}>{label}</label>)}
      <Component
        inputId={id || undefined}
        className='multipleSelect'
        classNamePrefix='multipleSelect'
        aria-label={ariaLabel}
        isDisabled={isDisabled}
        isMulti
        isLoading={isLoading}
        isSearchable={isSearchable}
        menuPortalTarget={menuPortalTarget || document.body}
        closeMenuOnSelect={false}
        value={values}
        options={options}
        placeholder=''
        components={{
          Option: (optionProps) => <MultipleOption<T> {...optionProps} {...customProps} />,
          MultiValueRemove: (optionProps) => <MultipleValueRemove {...optionProps} {...customProps} />,
          MultiValueLabel: (optionProps) => <MultipleValueLabel {...optionProps} {...customProps} />
        }}
        onChange={onSelectChange}
        hideSelectedOptions={hideSelectedOptions || false}
        styles={{
          container: (styles: any) => ({
            ...styles,
            marginTop: '8px',
            minHeight: size === 'small' ? '35px' : '48px'
          }),
          control: (styles: any) => ({
            ...styles,
            minHeight: size === 'small' ? '35px' : '48px',
            borderWidth: '1px',
            borderStyle: 'solid',
            borderColor: error ? 'var(--navds-select-color-border-error)' : 'var(--navds-select-color-border)',
            backgroundColor: isDisabled ? 'var(--navds-semantic-color-component-background-alternate)' : 'var(--navds-select-color-background)',
            borderRadius: '4px',
            color: 'var(--navds-color-text-primary)'
          }),
          indicatorSeparator: (styles: any) => ({
            ...styles,
            backgroundColor: 'var(--navds-semantic-color-border)'
          }),
          menu: (styles: any) => ({
            ...styles,
            zIndex: 500,
            width: 'max-content',
            minWidth: '100%'
          }),
          multiValue: (styles: any) => ({
            ...styles,
            borderRadius: '20px',
            borderWidth: '1px',
            borderColor: 'var(--navds-semantic-color-border)',
            borderStyle: 'solid',
            padding: '0rem 0.25rem',
            backgroundColor: 'var(--navds-select-color-background)',
            color: 'var(--navds-color-text-primary)',
            alignItems: 'center',
            flexDirection: 'row',
            justifyContent: 'center'
          }),
          multiValueRemove: (styles: any) => ({
            ...styles,
            padding: '0rem'
          }),
          menuList: (styles: any) => ({
            ...styles,
            borderWidth: '1px',
            borderColor: 'var(--navds-semantic-color-border)',
            borderStyle: 'solid',
            backgroundColor: 'var(--navds-semantic-color-component-background-alternate)'
          }),
          menuPortal: (base: any) => ({ ...base, zIndex: 9999 }),
          option: (styles: any, { isFocused, isSelected }: any) => ({
            ...styles,
            padding: '0.5rem',
            color: isFocused
              ? 'var(--navds-semantic-color-text-inverted)'
              : isSelected
                ? 'var(--navds-semantic-color-text-inverted)'
                : 'var(--navds-color-text-primary)',
            backgroundColor: isFocused
              ? 'var(--navds-semantic-color-focus)'
              : isSelected
                ? 'var(--navds-semantic-color-interaction-primary-selected)'
                : 'var(--navds-select-color-background)'
          }),
          singleValue: (styles: any) => ({
            ...styles,
            color: 'var(--navds-color-text-primary)'
          })
        }}
        tabSelectsValue={false}
      />

      {error && (
        <div style={{ marginTop: '8px' }} role='alert' aria-live='assertive' className='navds-error-message navds-error-message--medium navds-label'>
          {error}
        </div>
      )}
    </MultipleSelectDiv>
  )
}

MultipleSelect.propTypes = {
  ariaLabel: PT.string,
  className: PT.string,
  creatable: PT.bool,
  error: PT.string,
  hideSelectedOptions: PT.bool,
  id: PT.string,
  isLoading: PT.bool,
  label: PT.oneOfType([PT.element, PT.string]).isRequired,
  onSelect: PT.func,
  options: PT.array,
  values: PT.array
}
export default MultipleSelect
