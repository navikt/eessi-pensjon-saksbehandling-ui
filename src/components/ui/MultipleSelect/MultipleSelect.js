import React, { useState, useEffect } from 'react'
import Select from 'react-select'
import CreatableSelect from 'react-select/lib/Creatable'
import PT from 'prop-types'
import _ from 'lodash'
import classNames from 'classnames'
import MultipleOption from './MultipleOption'
import MultipleValueRemove from './MultipleValueRemove'
import './MultipleSelect.css'

const MultipleSelect = (props) => {
  const [ _values, setValues ] = useState([])
  const [ mounted, setMounted ] = useState(false)

  const {
    id,
    placeholder,
    values,
    creatable = false,
    includeList,
    excludeList,
    optionList = [],
    className,
    error = false,
    components,
    errorMessage,
    onChange,
    hideSelectedOptions
  } = props

  useEffect(() => {
    if (!mounted) {
      setValues(values)
      setMounted(true)
    }
  }, [values, mounted])

  const include = (selectedValues, allValues) => {
    return _.filter(allValues, it => {
      return selectedValues.indexOf(it.value) >= 0
    })
  }

  const exclude = (selectedValues, allValues) => {
    return _.filter(allValues, it => {
      return selectedValues.indexOf(it.value) < 0
    })
  }

  const _onChange = (args) => {
    setValues(args)
    if (typeof onChange === 'function') {
      onChange(args)
    }
  }

  const selectStyle = () => {
    return {
      container: (styles, state) => {
        let { boxShadow, ...rest } = styles
        return {
          ...rest,
          backgroundColor: error ? '#f3e3e3' : '#fff',
          borderRadius: 4,
          borderColor: error ? '1 px solid #ba3a26' : '20px solid #b7b1a9',
          boxShadow: state.isFocused ? '0 0 0 3px #254b6d' : ''
        }
      },
      control: (styles, state) => {
        let { boxShadow, ...rest } = styles
        return {
          ...rest,
          borderColor: error ? '#ba3a26' : '#b7b1a9',
          backgroundColor: error ? '#f3e3e3' : '#fff',
          ':hover': {
            borderColor: '#0067c5',
            transition: 'border-color 200ms cubic-bezier(0.465, 0.183, 0.153, 0.946)'
          }
        }
      }
    }
  }

  let options = includeList ? include(includeList, optionList) : optionList
  options = excludeList ? exclude(excludeList, options) : options

  return <div id={id} className={classNames('c-ui-multipleSelect', className, { 'skjemaelement__feilmelding': error })}>
    {creatable
      ? <CreatableSelect placeholder={placeholder}
        isMulti
        closeMenuOnSelect={false}
        defaultValue={_values}
        options={options}
        id={id ? id + '-select' : null}
        components={{
          Option: MultipleOption,
          MultiValueRemove: MultipleValueRemove,
          ...components }}
        className='multipleSelect'
        classNamePrefix='multipleSelect'
        onChange={_onChange}
        hideSelectedOptions={hideSelectedOptions || false}
        styles={selectStyle()}
        tabSelectsValue={false}
      />
      : <Select placeholder={placeholder}
        isMulti
        closeMenuOnSelect={false}
        defaultValue={_values}
        options={options}
        id={id ? id + '-select' : null}
        components={{
          Option: MultipleOption,
          MultiValueRemove: MultipleValueRemove,
          ...components }}
        className='multipleSelect'
        classNamePrefix='multipleSelect'
        onChange={_onChange}
        hideSelectedOptions={hideSelectedOptions || false}
        styles={selectStyle()}
        tabSelectsValue={false}
      />}
    {error
      ? <div role='alert' aria-live='assertive'>
        <div className='skjemaelement__feilmelding'>
          {errorMessage}
        </div>
      </div>
      : null
    }
  </div>
}

MultipleSelect.propTypes = {
  onChange: PT.func.isRequired,
  value: PT.oneOfType([PT.object, PT.string]),
  style: PT.object,
  includeList: PT.array,
  excludeList: PT.array,
  optionList: PT.array,
  className: PT.string,
  required: PT.string,
  id: PT.string,
  inputProps: PT.object,
  customInputProps: PT.object,
  errorMessage: PT.string,
  styles: PT.object,
  error: PT.bool,
  components: PT.object,
  creatable: PT.bool
}

export default MultipleSelect
