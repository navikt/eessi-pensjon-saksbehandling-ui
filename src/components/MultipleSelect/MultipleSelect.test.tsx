import { Option } from 'src/declarations/app.d'
import MultipleSelect, { MultipleSelectProps } from './MultipleSelect'
import {screen, render, fireEvent} from '@testing-library/react'

describe('components/MultipleSelect/MultipleSelect', () => {
  const options = [
    { label: 'mockLabel01', value: 'mockValue01' },
    { label: 'mockLabel02', value: 'mockValue02' }
  ] as Array<Option>

  const initialMockProps: MultipleSelectProps<Option> = {
    ariaLabel: 'mockAriaLabel',
    creatable: true,
    isDisabled: false,
    error: undefined,
    hideSelectedOptions: false,
    id: 'mockMultipleSelectId',
    isLoading: false,
    label: 'mockLabel',
    onSelect: jest.fn(),
    options,
    values: [options[1]]
  }

  it('Render: has proper HTML structure', () => {
    render(<MultipleSelect {...initialMockProps} />)
    expect(screen.getByTestId('c-multipleSelect')).toBeInTheDocument()
  })

  it('Handling: Triggers onSelect', () => {
    render(<MultipleSelect {...initialMockProps} />)
    fireEvent.keyDown(screen.getByRole('combobox'), { key: 'ArrowDown' })
    fireEvent.keyDown(screen.getByRole('combobox'), { key: 'Enter' })
    expect(initialMockProps.onSelect).toHaveBeenCalledWith([options[1], options[0]])
  })
})
