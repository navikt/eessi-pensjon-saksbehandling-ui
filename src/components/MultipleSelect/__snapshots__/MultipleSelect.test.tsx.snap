// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MultipleSelect/MultipleSelect Render: match snapshot 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

ui-fss.g1[id="c2"] {
  content: "c3,";
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
}

.c5:hover {
  background-color: transparent !important;
}

ui-fss.g2[id="c4"] {
  content: "c5,";
}

.c1 .skjemaelement--feilmelding .multipleSelect--control {
  border: 1px solid;
}

ui-fss.g4[id="c0"] {
  content: "c1,";
}

<div
  class="c0 c1"
>
  <label
    class="navds-text-field--label navds-label"
    for="mockMultipleSelectId"
  >
    mockLabel
  </label>
  <div
    class="multipleSelect css-91gbow-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class="multipleSelect__control css-1q6azai-control"
    >
      <div
        class="multipleSelect__value-container multipleSelect__value-container--is-multi multipleSelect__value-container--has-value css-g1d714-ValueContainer"
      >
        <div
          class="css-drugbm-multiValue multipleSelect__multi-value"
        >
          <div
            class="css-12jo7m5 multipleSelect__multi-value__label"
          >
            <div
              class="c2 c3"
            >
              <p
                class="navds-body-long"
              >
                mockLabel02
              </p>
            </div>
          </div>
          <div
            aria-label="Remove mockLabel02"
            class="c4 c5 css-xscnok multipleSelect__multi-value__remove"
            data-testid="c-multipleselect-multiplevalueremove"
          >
            <svg
              fill="none"
              focusable="false"
              height="1em"
              role="img"
              style="visibility: visible;"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12 0c6.627 0 12 5.373 12 12s-5.373 12-12 12S0 18.627 0 12 5.373 0 12 0Zm3.571 7L17 8.429 13.428 12 17 15.571 15.571 17 12 13.428 8.429 17 7 15.571 10.572 12 7 8.429 8.429 7 12 10.572 15.571 7Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="multipleSelect__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="mockAriaLabel"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="multipleSelect__input"
            id="mockMultipleSelectId"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="multipleSelect__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="multipleSelect__indicator multipleSelect__clear-indicator css-tlfecz-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="multipleSelect__indicator-separator css-141yg44-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="multipleSelect__indicator multipleSelect__dropdown-indicator css-tlfecz-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
